log/levels/levels.go

```diff
@@ -29,8 +29,54 @@ type Levels struct {
 	warnValue  string
 	errorValue string
 	critValue  string
+
+	// FilterLogLevel is an option allowing logs below a certain level to be
+	// discarded rather than logged. For example, a consumer may set LogLevel
+	// to levels.Warn, and any logs to Debug or Info would become no-ops.
+	filterLogLevel LogLevel
+
+	// Promote errors
+	promoteErrors       bool
+	errorKey            string
+	promoteErrorToLevel LogLevel
+}
+
+// levelCommittedLogger embeds a log level that the user has selected and prepends
+// the associated level string as a prefix to every log line. Log level may be
+// affected by error promotion; and log lines may be suppressed based on the value
+// of FilterLogLevel.
+type levelCommittedLogger struct {
```

Comment:
To be more verbose...

The `Logger` interface is very short:

    type Logger interface {
    	Log(keyvals ...interface{}) error
    }

`levelCommittedLogger` implements the `Log` method at line 130 in this file, so it satisfies the `log.Logger` interface.

Let's look specifically at the `levels.Debug` case. Previously it returned a `log.Logger`, and it still does. Before this change, it returned `l.ctx.WithPrefix(l.levelKey, l.debugValue)`

With this change, it returns a `levelCommittedLogger{l, Debug}`.  Whose `Log` method becomes:

    lvl, ctx := l.committedLevel, l.levels.ctx.With(keyvals...)
    //...
    if l.levels.promoteErrors {...}  // By default, l.levels.promoteErrors is false, so short circuit this block
    //...
    if lvl < l.levels.filterLogLevel {...} // By default l.levels.filerLogLevel is 0 and lvl ranges from 0 to 4,
                                           // so short circuit this block
    //...
    var levelValue string
    switch lvl {
    case Debug:
    	levelValue = l.levels.debugValue
    	// Short circuit the rest of the cases, because case Debug is true
    }
    //...
    return ctx.WithPrefix(l.levels.levelKey, levelValue).Log()

In other words, with appropriate substitutions, that last line becomes:

    return l.levels.ctx.With(keyvals...).WithPrefix(l.levels.levelKey, l.levels.debugValue).Log()

The only difference between this line and

    l.ctx.WithPrefix(l.levelKey, l.debugValue).Log(keyvals...)

...is that you can't overwrite `l.levelKey` with a different ad-hoc value, which you shouldn't be doing anyway if you're using the `levels` package.